### PR TITLE
Prohibit and remove unused declarations in the processing module

### PR DESCRIPTION
**Pull request by leventov**
I started doing this in the whole project, but quickly realised that it would be a too big PR and only enforced and completed in the processing module.

### CODE REVIEW

File: api/src/main/java/io/druid/cli/CliRunnable.java [hunk 1]
```diff
@@ -1,26 +0,0 @@
...
- * KIND, either express or implied. See the License for the
- * specific language governing permissions and limitations
- * under the License.
- */
-
-package io.druid.cli;
-
-/**
- */
-public interface CliRunnable extends Runnable
```
Comment (drcrallen) [inline api/src/main/java/io/druid/cli/CliRunnable.java]: You can't remove extension points.

**Comment by leventov** [comment]
@jihoonson could you please review?

File: extensions-contrib/azure-extensions/src/main/java/io/druid/firehose/azure/AzureBlob.java [hunk 2]
```diff
@@ -35,10 +36,7 @@
   @NotNull
   private String path = null;
 
-  public AzureBlob()
```
Comment (jihoonson) [inline extensions-contrib/azure-extensions/src/main/java/io/druid/firehose/azure/AzureBlob.java]: This is needed for JSON serde.

File: processing/src/main/java/io/druid/segment/MetricHolder.java [hunk 3]
```diff
@@ -155,24 +75,14 @@ public static MetricHolder fromByteBuffer(ByteBuffer buf, ObjectStrategy strateg
           holder.complexType = GenericIndexed.read(buf, serdeForType.getObjectStrategy());
         }
         break;
+      case LONG:
+      case DOUBLE:
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/MetricHolder.java]: Why this change?

Comment (leventov) [reply processing/src/main/java/io/druid/segment/MetricHolder.java]: Apparently MetricHolder is used only when dealing with legacy segment format, in which long- and double-typed columns doesn't exist

File: extensions-contrib/azure-extensions/src/main/java/io/druid/firehose/azure/AzureBlob.java (hunk 2 above)
Comment (leventov) [reply extensions-contrib/azure-extensions/src/main/java/io/druid/firehose/azure/AzureBlob.java]: Fixed

**Comment by leventov** [comment]
@jihoonson thanks, addressed comments

**Comment by leventov** [comment]
@jihoonson more comments here?

File: processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java [hunk 4]
```diff
@@ -118,7 +96,7 @@ public ImmutableBitmap fromByteBuffer(ByteBuffer buffer, int numBytes)
     @Override
     public int compare(ImmutableBitmap o1, ImmutableBitmap o2)
     {
-      return RORING_COMPARATOR.compare((WrappedImmutableRoaringBitmap) o1, (WrappedImmutableRoaringBitmap) o2);
+      throw new UnsupportedOperationException();
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java]: Should this be removed? I think it's useful and might be used for index-only query processing in the future. 
If this should be removed, it's better to make ObjectStrategy to not extend Comparator.

File: processing/src/test/java/io/druid/query/aggregation/JavaScriptAggregatorBenchmark.java [hunk 5]
```diff
@@ -26,6 +26,10 @@
 import java.util.Collections;
 import java.util.Map;
 
+
+/**
+ * TODO rewrite to use JMH and move to benchmarks project
```
Comment (jihoonson) [inline processing/src/test/java/io/druid/query/aggregation/JavaScriptAggregatorBenchmark.java]: Would you open a new issue for this?

File: processing/src/test/java/io/druid/query/aggregation/cardinality/CardinalityAggregatorBenchmark.java [hunk 6]
```diff
@@ -38,6 +38,9 @@
 import java.nio.ByteBuffer;
 import java.util.List;
 
+/**
+ * TODO rewrite to use JMH and move to the benchmarks project
```
Comment (jihoonson) [inline processing/src/test/java/io/druid/query/aggregation/cardinality/CardinalityAggregatorBenchmark.java]: Would you open a new issue for this?

File: processing/src/test/java/io/druid/query/topn/TopNBinaryFnBenchmark.java [hunk 7]
```diff
@@ -41,6 +41,9 @@
 import java.util.List;
 import java.util.Map;
 
+/**
+ * TODO rewrite to use JMH and move to the benchmarks project
```
Comment (jihoonson) [inline processing/src/test/java/io/druid/query/topn/TopNBinaryFnBenchmark.java]: Would you open a new issue for this?

File: bytebuffer-collections/src/test/java/io/druid/collections/spatial/ImmutableRTreeTest.java [hunk 8]
```diff
@@ -545,7 +544,7 @@ public void testSearchWithSplitLimitedBoundRoaring()
     }
   }
 
-  //@Test
+  @SuppressWarnings("unused") // TODO rewrite to JMH and move to the benchmarks project
```
[the remark this answers is not in the document]
Comment (leventov) [reply bytebuffer-collections/src/test/java/io/druid/collections/spatial/ImmutableRTreeTest.java]: https://github.com/druid-io/druid/issues/5030

File: .idea/misc.xml [hunk 9]
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (leventov) [reply .idea/misc.xml]: Unfortunately this couldn't be renamed, there must be only one misc.xml in IntelliJ project, as far as I know. This is silly and cause many complains (you can google "misc xml git"), but there is no other way. What plugins do you use and what kind of info they store in this file?

File: .idea/misc.xml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        <list size="4">
+          <item index="0" class="java.lang.String" itemvalue="org.jetbrains.annotations.NotNull" />
+          <item index="1" class="java.lang.String" itemvalue="javax.annotation.Nonnull" />
+          <item index="2" class="java.lang.String" itemvalue="edu.umd.cs.findbugs.annotations.NonNull" />
+          <item index="3" class="java.lang.String" itemvalue="android.support.annotation.NonNull" />
+        </list>
+      </value>
+    </option>
+  </component>
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="false" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
[the remark this answers is not in the document]
Comment (leventov) [reply .idea/misc.xml]: To avoid conflicts, developers will need to rename their JDK to "1.8", it could be done in IntelliJ in "Project Structure..." -> "Platform Settings" -> "SDKs".

File: bytebuffer-collections/src/main/java/io/druid/collections/bitmap/MutableBitmap.java [hunk 11]
```diff
@@ -41,36 +39,6 @@
...
 
-  /**
-   * Compute the bitwise-and of this bitmap with another bitmap. The current
-   * bitmap is modified whereas the other bitmap is left intact.
-   *
-   * Note that the other bitmap should be of the same class instance.
-   *
-   * @param mutableBitmap other bitmap
-   */
-  void and(MutableBitmap mutableBitmap);
```
[the remark this answers is not in the document]
Comment (leventov) [reply bytebuffer-collections/src/main/java/io/druid/collections/bitmap/MutableBitmap.java]: When it is needed in the future, it could be added back

File: processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java (hunk 4 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java]: It was used because Bitmaps are serialized using GenericIndexedWriter, which checks if objects are sorted. It makes sense for e. g. Strings, but not bitmaps, that is pointless. So I added `setObjectsNotSorted()` method to `GenericIndexedWriter` to avoid doing pointless work when serializing things such as bitmaps.

File: bytebuffer-collections/src/main/java/io/druid/collections/bitmap/ImmutableBitmap.java [hunk 12]
```diff
@@ -39,8 +39,6 @@
 
   byte[] toBytes();
 
-  int compareTo(ImmutableBitmap other);
```
[the remark this answers is not in the document]
Comment (leventov) [reply bytebuffer-collections/src/main/java/io/druid/collections/bitmap/ImmutableBitmap.java]: Bitmap ordering doesn't make any sense to me, I don't see how it could be used in the future.

File: .idea/misc.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply .idea/misc.xml]: The conflict occurs because this file is added to git in this patch, but it already exists in my machine. 

> Unfortunately this couldn't be renamed, there must be only one misc.xml in IntelliJ project, as far as I know.

Then, I would like to be conservative about this issue. People may want to add their own options to this file which might be difficult to be shared and thus maintained by git.

Is it possible to move EntryPointsManager configuration to another file? I think it's the best option. If not possible, I suggest to not add it.

File: .idea/misc.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply .idea/misc.xml]: Hmm, I don't think fixing the project-jdk-name is a good idea. It's possible that some developers have installed two or more JDKs in their machines. Suggest to remove this.

File: bytebuffer-collections/src/main/java/io/druid/collections/bitmap/MutableBitmap.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply bytebuffer-collections/src/main/java/io/druid/collections/bitmap/MutableBitmap.java]: Ok.

File: indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java [hunk 13]
```diff
@@ -135,26 +134,19 @@
   }
 
   private final HadoopDruidIndexerConfig config;
-  private IndexGeneratorStats jobStats;
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java]: Please address this comment.

File: processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java (hunk 4 above)
Comment (jihoonson) [reply processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java]: I see. Then, ConciseBitmapSerdeFactory has the same issue.

File: .idea/misc.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (leventov) [reply .idea/misc.xml]: > The conflict occurs because this file is added to git in this patch, but it already exists in my machine.

Yes, but once this file is checked in the repo, there shouldn't be more conflicts.

> People may want to add their own options to this file which might be difficult to be shared and thus maintained by git.

Note that this is per-project settings. The trade-off is to ask people not to use other per-project settings, when working specifically with the druid project, or else "share" them in PRs.

> Is it possible to move EntryPointsManager configuration to another file? I think it's the best option. If not possible, I suggest to not add it.

It's not possible. Not including it will not allow to track unused declarations using IntelliJ inspections CI.

File: .idea/misc.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (leventov) [reply .idea/misc.xml]: > It's possible that some developers have installed two or more JDKs in their machines.

I don't see why multiple JDKs is the problem.

> Suggest to remove this.

The problem is that IntelliJ will add this line automatically in the file whenever it can, so you will always have "unchecked" changes in misc.xml, from Git point of view.

File: processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java (hunk 4 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java]: Fixed

File: indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (leventov) [reply indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java]: Restored jobStats, added a comment that they are not used at the moment.

File: .idea/misc.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply .idea/misc.xml]: > I don't see why multiple JDKs is the problem.

JDK configurations are shared across multiple projects, and multiple JDKs should have different names which are usually intuitive like ```1.8 oracle```. If there is no JDK of name ```1.8```, Intellij will fail to find a JDK automatically. Finally, developers should set one of their JDKs as the project SDK which modifies this file, or rename one of their JDKs to ```1.8```. I think both cases are not good.

File: .idea/misc.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply .idea/misc.xml]: > The trade-off is to ask people not to use other per-project settings, when working specifically with the druid project, or else "share" them in PRs.

Hmm, I basically agree on that this default configuration can be useful for developers. However, I think it shouldn't be mandatory. This is because 1) this configuration is specific to only Intellij, so developers using other tools cannot benefit from this and 2) IDE-specific configurations should be able to be changed easily without worrying about PR review or code maintenance (like the SDK name issue below). 

It may be fine if this file is rarely changed, but I'm not sure which kind of configurations it can contain. Do you have any links to any document about this file? I couldn't find anything.

Maybe it's better to maintain IDE-specific configurations in some other places and let developers use or copy them. This can be done by documentation.

File: .idea/misc.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (leventov) [reply .idea/misc.xml]: Developers could also add "another" JDK which is called `1.8` and points to the same files as some existing JDK, even if it's 1.9 in fact. `1.8` is just an alias name. I agree that it's not ideal, but I don't see a better decision

File: .idea/misc.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (leventov) [reply .idea/misc.xml]: This official answer suggests to check into VCS all files under `.idea` dir, with few exceptions, but `misc.xml` is not one of those, so it should be included: https://intellij-support.jetbrains.com/hc/en-us/articles/206544839-How-to-manage-projects-under-Version-Control-Systems

Regarding other IDEs: I don't think we used to try to keep Eclipse and Idea settings/configs in sync, and Eclipse is lagging now, because less developers use it. Partially to mitigate this, IntelliJ inspections CI was introduced, and _this_ is why I'm checking misc.xml now.

File: .idea/misc.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (leventov) [reply .idea/misc.xml]: Added a section to INTELLIJ_SETUP.md

**Comment by leventov** [comment]
@jihoonson could you please take another look?

File: .idea/misc.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply .idea/misc.xml]: I see. The only thing I'm concerned with is frequent conflicts due to different per-developer settings, but it looks that per-developer settings are not stored in this file except the project SDK name. It sounds we can compromise for more efficient development and code review.

File: .idea/misc.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply .idea/misc.xml]: Thanks.

**Comment by leventov** [comment]
@jihoonson could you please merge this PR?

**Comment by gianm** [comment]
I'll merge it, it seems clear that @jihoonson approves.

**Comment by leventov** [comment]
@gianm thanks
